readthedocs/core/views/hooks.py

```diff
@@ -136,91 +137,135 @@ def _build_url(url, projects, branches):
         except NoProjectException:
-            log.error(
-                "(Incoming GitHub Build) Repo not found:  %s" % hacked_repo_url)
-            return HttpResponseNotFound('Repo not found: %s' % hacked_repo_url)
+            log.error('Project match not found: url=%s', http_search_url)
+            return HttpResponseNotFound('Project not found')
     else:
-        return HttpResponse("You must POST to this resource.")
+        return HttpResponse('Method not allowed', status=405)
 
 
 @csrf_exempt
 def gitlab_build(request):
-    """A post-commit hook for GitLab."""
+    """GitLab webhook consumer
+
+    Search project repository URLs using the site URL from GitLab webhook payload.
+    This search is error-prone, use the API v2 webhook view for new webhooks.
+    """
     if request.method == 'POST':
         try:
-            # GitLab RTD integration
-            obj = json.loads(request.POST['payload'])
-        except:
-            # Generic post-commit hook
-            obj = json.loads(request.body)
-        url = obj['repository']['homepage']
-        ghetto_url = url.replace('http://', '').replace('https://', '')
-        branch = obj['ref'].replace('refs/heads/', '')
-        log.info("(Incoming GitLab Build) %s [%s]" % (ghetto_url, branch))
-        projects = get_project_from_url(ghetto_url)
+            data = json.loads(request.body)
+        except (ValueError, TypeError):
+            log.error('Invalid GitLab webhook payload', exc_info=True)
+            return HttpResponse('Invalid request', status=400)
+        url = data['project']['http_url']
+        search_url = re.sub(r'^https?://(.*?)(?:\.git|)$', '\\1', url)
+        branches = [data['ref'].replace('refs/heads/', '')]
+        log.info(
+            'GitLab webhook search: url=%s branches=%s',
+            search_url,
+            branches
+        )
+        projects = get_project_from_url(search_url)
         if projects:
-            return _build_url(ghetto_url, projects, [branch])
+            return _build_url(search_url, projects, branches)
         else:
-            log.error(
-                "(Incoming GitLab Build) Repo not found:  %s" % ghetto_url)
-            return HttpResponseNotFound('Repo not found: %s' % ghetto_url)
+            log.error('Project match not found: url=%s', search_url)
+            return HttpResponseNotFound('Project match not found')
     else:
-        return HttpResponse("You must POST to this resource.")
+        return HttpResponse('Method not allowed', status=405)
 
 
 @csrf_exempt
 def bitbucket_build(request):
+    """Consume webhooks from multiple versions of Bitbucket's API
+
+    API v1
+        https://confluence.atlassian.com/bitbucket/events-resources-296095220.html
+
+    API v2
+        https://confluence.atlassian.com/bitbucket/event-payloads-740262817.html#EventPayloads-Push
+    """
     if request.method == 'POST':
-        payload = request.POST.get('payload')
```

Comment:
Think this is similarly required for old BB API's no? We're depending on request.body being JSON here, but presumably this code was working previously, so some BB requests were using this method?